feat: add documentation for setting up GCP and GKE (#75)
charmed-hpc/docs: howto/initialize-cloud-environment.md

:::

[regions]: https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Concepts.RegionsAndAvailabilityZones.html

::::

::::{tab-item} Google Cloud Platform (GCP)
:sync: gcp

> :::{warning}
> Currently, Juju does not support any kind of isolation between Juju models created in a GCP project
> and other resources in the same project. This could make the cluster difficult to manage if you have
> special security requirements.
> :::

Reviewer: ```suggestion
:::{warning}
Currently, Juju does not support any kind of isolation between Juju models and other resources created within the same GCP project. This may make some processes more complex if you have special security requirements.
:::
```

To use GCP as the machine cloud for your Charmed HPC cluster, you will need to have:

* [Installed the gcloud CLI](https://cloud.google.com/sdk/docs/install)
* [Authenticated into the gcloud CLI](https://cloud.google.com/docs/authentication/gcloud#local)
* [Adjusted quotas for suitable Compute Engine machine families](https://cloud.google.com/compute/resource-usage#cpu_quota)

To decide on suitable instance types, it may be useful to refer to GCP's [Machine families resource and comparison guide][machines].
A typical Charmed HPC deployment will likely use a mix of Compute-optimized and Accelerator-optimized VMs
for cluster compute nodes, and General-purpose VMs for other node types.

[machines]: https://cloud.google.com/compute/docs/machine-resource

### Create a GCP Service Account

Juju requires a JSON file containing the Service Account key to authenticate Juju, which allows

> it to create, remove and manage machines.

Reviewer: ```suggestion
it to create, remove, and manage machines.
```

:::{code-block} shell
gcloud projects add-iam-policy-binding my-project \
  --member="serviceAccount:JujuService@my-project.iam.gserviceaccount.com" \
  --role="roles/compute.instanceAdmin.v1" \
  --role="roles/compute.securityAdmin"
:::

After creating the required Service Account, generate the Service Account key file for the `JujuService`
user using the following command:

:::{code-block} shell
gcloud iam service-accounts keys create ~/.local/share/juju/juju-sa-creds.json
:::

:::{note}
It is very important that the key file is located in the directory path `~/.local/share/juju` for the
user that will manage the Juju CLI, since the tool does not have permissions to read files in other
directories.
:::

### Add GCP cloud credentials to Juju

To make your GCP credentials known to Juju, run:

:::{code-block} shell
juju add-credential google
:::

This will start a set of prompts where you will be asked:

* `credential-name` — Your choice of name that will help you identify the credential set, e.g. `my-gcp-credential`.
* `region` — The region the credential is tied to e.g. `us-east-1`.
  Since GCP users are not tied to a specific GCP region, this can be left blank.

> * `auth type` -- The method of authentication. Select `jsonfile` here.

Reviewer: Is this a 'type it in' type prompt, or a 'choose from a list and select the number' type, or something else? Using "Select" as the verb here might not be ideal depending on how the prompt actually works.

Author: Funny thing about this prompt is that you need to choose an option from a list, but you also need to type out the option 😅

Author: Reworded this part a bit to reflect that you need to input the option name.

> * `Path` -- The path to the Service Account key file, which corresponds to `~/.local/share/juju/juju-sa-creds.json`.

Author: I _think_ the CLI also accepts relative paths, but it's safer to put the full path so I'll commit your suggestion instead.
